e2e/test/scenarios/joins/joins-reproductions.cy.spec.js

```diff
@@ -1262,4 +1262,28 @@ describe("issue 42385", { tags: "@external" }, () => {
       cy.findByLabelText("Right column").should("not.exist");
     });
   });
+
+  it("should remove invalid join clause in incomplete draft state when query database changes (metabase#42385)", () => {
+    openOrdersTable({ mode: "notebook" });
+    join();
+    entityPickerModal().within(() => {
+      entityPickerModalTab("Tables").click();
+      cy.findByText("Products").click();
+    });
```

Comment:
At this point we have a valid, complete, non-draft join clause: [image](https://github.com/metabase/metabase/assets/6830683/c7fd8893-2054-485a-ad31-3ae95789fb17)